### PR TITLE
feat(redis_client): basic `RedisClient`

### PR DESCRIPTION
**Pull request by felangel**
## Description

- feat(redis_client): basic `RedisClient` API

```dart
import 'dart:async';

import 'package:redis_client/redis_client.dart';

Future<void> main() async {
  // Create an instance of a RedisClient.
  final client = RedisClient();

  // Connect to the Redis server.
  await client.connect();

  // Set the value of a key.
  await client.set(key: 'HELLO', value: 'WORLD');

  // Get the value of a key.
  final value = await client.get(key: 'HELLO'); // WORLD

  // Delete the key.
  await client.delete(key: 'HELLO');

  // Close the connection to the Redis server.
  await client.close();
}
```

## Type of Change

- [X] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: packages/redis_client/README.md [hunk 1]
```diff
@@ -7,6 +7,34 @@ A Dart library for interacting with [Redis](https://redis.io).
...
+import 'dart:async';
+
+import 'package:redis_client/redis_client.dart';
+
+Future<void> main() async {
+  // Create an instance of a RedisClient.
+  final client = RedisClient();
+
+  // Connect to the Redis server.
+  await client.connect();
```
Comment (bryanoltman) [inline packages/redis_client/README.md:22]: At what point is the server URL/other connection info specified?

File: packages/redis_client/lib/src/redis_client.dart [hunk 2]
```diff
@@ -1,7 +1,241 @@
...
+
+/// {@template redis_socket_options}
+/// Options for connecting to a Redis server.
+/// {@endtemplate}
+class RedisSocketOptions {
+  /// {@macro redis_socket_options}
+  const RedisSocketOptions({
+    this.host = 'localhost',
+    this.port = 6379,
+    this.timeout = const Duration(seconds: 30),
```
Comment (bryanoltman) [inline packages/redis_client/lib/src/redis_client.dart:16]: Ah! Here it is. Is 6379 the default Redis port?

Comment (bryanoltman) [reply packages/redis_client/lib/src/redis_client.dart:16]: Generally, I'd update the doc comments with notes about the defaults.

File: packages/redis_client/test/src/redis_client_test.dart [hunk 3]
```diff
@@ -1,10 +1,157 @@
...
+        client.connect(),
+        throwsA(
+          isA<StateError>().having(
+            (e) => e.message,
+            'message',
+            'RedisClient has been closed.',
+          ),
+        ),
+      );
+    });
```
Comment (bryanoltman) [inline packages/redis_client/test/src/redis_client_test.dart]: I'd consider putting these in a "connection" group

File: packages/redis_client/test/src/redis_client_test.dart [hunk 4]
```diff
@@ -1,10 +1,157 @@
...
+            ),
+          ),
+        );
+      });
+
+      test('fails when username is incorrect', () async {
+        await expectLater(
+          client.auth(username: 'shorebird', password: 'password'),
+          completion(equals(false)),
+        );
```
Comment (bryanoltman) [inline packages/redis_client/test/src/redis_client_test.dart]: Stylistic question: is there an advantage to this form of `expect` over `expect(await myFunction(), isTrue)`?

Comment (bryanoltman) [reply packages/redis_client/test/src/redis_client_test.dart]: Either way, `isTrue`/`isFalse` is probably more idiomatic use of the matcher framework

**Review by bryanoltman — APPROVED** [review]
A few small nits, otherwise LGTM!

File: packages/redis_client/README.md (hunk 1 above)
Comment (felangel) [reply packages/redis_client/README.md:22]: It's passed as `RedisSocketOptions` via the constructor but there are defaults preconfigured.

File: packages/redis_client/lib/src/redis_client.dart (hunk 2 above)
Comment (felangel) [reply packages/redis_client/lib/src/redis_client.dart:16]: > Ah! Here it is. Is 6379 the default Redis port?

Yup these are all defaults. I'll update docs.

File: packages/redis_client/test/src/redis_client_test.dart (hunk 3 above)
Comment (felangel) [reply packages/redis_client/test/src/redis_client_test.dart]: yup good call 👍

File: packages/redis_client/test/src/redis_client_test.dart (hunk 4 above)
Comment (felangel) [reply packages/redis_client/test/src/redis_client_test.dart]: > Stylistic question: is there an advantage to this form of `expect` over `expect(await myFunction(), isTrue)`?

I believe the matcher description will be nicer if the test fails with the former as opposed to the later but it's mostly habitual 😆

**Pull request merged**